Importing pandas

In [1]:
import pandas as pd

Uploading the dataset, df

In [3]:
df = pd.read_csv('Advanced_pandas_sales_dataset.csv')
df.head()

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
0,1001,2024-10-01,Priya,Hyderabad,East,Mouse,Accessories,1735,7,0.00,307.52
1,1002,2024-05-03,Sneha,Kolkata,East,Monitor,Electronics,807,2,0.15,49.89
2,1003,2024-01-26,Anita,Mumbai,West,Phone,Electronics,2328,1,0.20,66.47
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30


Question 1:  Dataset Exploration

          Perform the following exploratory steps:
               Display the first 10 rows
               Display the dataset structure using info()
               Find the number of rows and columns
               Identify the data types of each column

In [4]:
# Displaying the first 10 rows of the dataset 
df.head(10)

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
0,1001,2024-10-01,Priya,Hyderabad,East,Mouse,Accessories,1735,7,0.00,307.52
1,1002,2024-05-03,Sneha,Kolkata,East,Monitor,Electronics,807,2,0.15,49.89
2,1003,2024-01-26,Anita,Mumbai,West,Phone,Electronics,2328,1,0.20,66.47
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30
5,1006,2024-07-04,Pooja,Bangalore,West,Printer,Office,3073,8,0.20,191.58
6,1007,2024-06-24,Dev,Chennai,East,Monitor,Electronics,1170,5,0.20,22.83
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
9,1010,2024-12-25,Pooja,Ahmedabad,North,Phone,Electronics,2711,8,0.00,477.32


In [5]:
# alternatively 
df.iloc[:10]

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
0,1001,2024-10-01,Priya,Hyderabad,East,Mouse,Accessories,1735,7,0.00,307.52
1,1002,2024-05-03,Sneha,Kolkata,East,Monitor,Electronics,807,2,0.15,49.89
2,1003,2024-01-26,Anita,Mumbai,West,Phone,Electronics,2328,1,0.20,66.47
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30
5,1006,2024-07-04,Pooja,Bangalore,West,Printer,Office,3073,8,0.20,191.58
6,1007,2024-06-24,Dev,Chennai,East,Monitor,Electronics,1170,5,0.20,22.83
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
9,1010,2024-12-25,Pooja,Ahmedabad,North,Phone,Electronics,2711,8,0.00,477.32


In [6]:
#  Displaying the dataset structure using info() 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 230 entries, 0 to 229
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Order_ID    230 non-null    int64  
 1   Order_Date  230 non-null    object 
 2   Customer    230 non-null    object 
 3   City        230 non-null    object 
 4   Region      230 non-null    object 
 5   Product     230 non-null    object 
 6   Category    230 non-null    object 
 7   Sales       230 non-null    int64  
 8   Quantity    230 non-null    int64  
 9   Discount    230 non-null    float64
 10  Profit      230 non-null    float64
dtypes: float64(2), int64(3), object(6)
memory usage: 19.9+ KB


In [7]:
# Finding the number of rows and columns   
df.shape

(230, 11)

In [9]:
# Identify the data types of each column 
df.dtypes

Order_ID        int64
Order_Date     object
Customer       object
City           object
Region         object
Product        object
Category       object
Sales           int64
Quantity        int64
Discount      float64
Profit        float64
dtype: object

Question 2: Sales Filtering Analysis

          Filter the dataset to display:

             Orders where Sales > 3500
             Orders where Region = "East" and Sales > 2500
             Explain what insights you observe.

In [10]:
#  Orders where Sales > 3500 
df[df['Sales'] > 3500]

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
10,1011,2024-06-30,Neha,Delhi,South,Printer,Office,5977,6,0.00,1578.32
...,...,...,...,...,...,...,...,...,...,...,...
219,1220,2024-03-22,Meera,Bangalore,West,Phone,Electronics,5986,7,0.05,986.70
222,1223,2024-03-04,Meera,Ahmedabad,South,Phone,Electronics,4356,6,0.10,416.87
223,1224,2024-01-07,Dev,Bangalore,East,Tablet,Electronics,4115,4,0.05,870.03
225,1226,2024-06-14,Anita,Ahmedabad,East,Laptop,Electronics,5647,2,0.15,662.00


In [11]:
#  Orders where Region = "East" and Sales > 2500

df[(df['Region']=='East')  & (df['Sales']> 2500)]

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
13,1014,2024-03-18,Anita,Bangalore,East,Mouse,Accessories,3616,4,0.05,463.93
21,1022,2024-01-14,Nitin,Chennai,East,Monitor,Electronics,4827,6,0.05,912.34
27,1028,2024-07-21,Dev,Kolkata,East,Tablet,Electronics,4054,6,0.00,1013.83
33,1034,2024-12-15,Nitin,Ahmedabad,East,Tablet,Electronics,4378,2,0.20,-104.30
34,1035,2024-02-20,Nitin,Ahmedabad,East,Laptop,Electronics,2535,4,0.10,188.15
41,1042,2024-07-15,Rohan,Chennai,East,Laptop,Electronics,3268,8,0.15,295.15
43,1044,2024-09-20,Kavya,Hyderabad,East,Printer,Office,2618,7,0.05,454.64


 Explain what insights you observe. 

   Explanation: 
   
   Some orders for example Order ID 1004 compared to other orders, has higher sales of 3933 but it is in negative profits. The same also                  applied to order 1034 which also has negative profits. 
   This is most probably arising from products being purchased expensively by the business and then sold out to the customers at a loss.
   There is need to outsource for products from suppliers who offer them at good rates to avoid having to sell items at a loss.
   

Question 3: Customer Purchase Analysis

          Identify the top 5 customers based on total sales.
          Hint: Use GroupBy with aggregation.

In [15]:
Top_5_customers = df.groupby('Customer')['Sales'].sum().nlargest(5)

In [16]:
print(Top_5_customers)

Customer
Nitin    59820
Raj      58416
Rahul    52291
Dev      48500
Isha     46818
Name: Sales, dtype: int64


Question 4: Regional Sales Performance
 
          Calculate the total sales generated by each region.
          Then identify:
             Which region generated the highest revenue?

In [18]:
# total sales generated by each region 
total_sales_by_region = df.groupby('Region')['Sales'].sum() 
print(total_sales_by_region)

Region
East     232675
North    158210
South    183108
West     170227
Name: Sales, dtype: int64


In [21]:
# region with the highest revenue 
highest_revenue_by_region = df.groupby('Region')['Sales'].sum().nlargest(1) 
print(highest_revenue_by_region)

Region
East    232675
Name: Sales, dtype: int64


Question 5: Product Performance Analysis

          Calculate the total quantity sold for each product.
          Which product appears to be the most popular?

# total quantity sold per product 

total_qty_per_product = df.groupby('Product')['Quantity'].sum()
print(total_qty_per_product)

In [25]:
# product that appears to be the most popular 

most_popular_product = df.groupby('Product')['Quantity'].sum().nlargest(1) 
print(most_popular_product)

Product
Monitor    156
Name: Quantity, dtype: int64


Question 6: Multi-Level Grouping

           Group the dataset by: Region and Category
           Calculate the total sales for each combination.
           What insights can you derive from this analysis?

In [34]:
# grouping the dataset by region and category 
df.groupby(['Region','Category']).head()

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
0,1001,2024-10-01,Priya,Hyderabad,East,Mouse,Accessories,1735,7,0.00,307.52
1,1002,2024-05-03,Sneha,Kolkata,East,Monitor,Electronics,807,2,0.15,49.89
2,1003,2024-01-26,Anita,Mumbai,West,Phone,Electronics,2328,1,0.20,66.47
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30
5,1006,2024-07-04,Pooja,Bangalore,West,Printer,Office,3073,8,0.20,191.58
6,1007,2024-06-24,Dev,Chennai,East,Monitor,Electronics,1170,5,0.20,22.83
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
9,1010,2024-12-25,Pooja,Ahmedabad,North,Phone,Electronics,2711,8,0.00,477.32


In [38]:
#  total sales for each combination 
df.groupby(['Region','Category'])['Sales'].sum()

Region  Category   
East    Accessories     57646
        Electronics    160907
        Office          14122
North   Accessories     62052
        Electronics     70023
        Office          26135
South   Accessories     80914
        Electronics     86690
        Office          15504
West    Accessories     64811
        Electronics     84128
        Office          21288
Name: Sales, dtype: int64

Insights derived from this analysis: 

    The most sales come from the East Region.
    
    The sale of electronics generate the most sales
    
    The sale of office stuff generates least sales

Question 7: Pivot Table Creation
    
         Create a pivot table showing:
            Rows → Region
            Columns → Product
            Values → Total Sales
         Which region sells the most laptops?

In [41]:
# pivot table 
df.pivot_table(values='Sales',columns='Product',index='Region',aggfunc='sum')

Product,Headphones,Keyboard,Laptop,Monitor,Mouse,Phone,Printer,Tablet
Region,,,,,,,,
East,23899,10845,55568,31299,22902,39667,14122,34373
North,15450,20538,4540,14922,26064,30535,26135,20026
South,24111,23480,23029,19167,33323,18218,15504,26276
West,17159,38499,10826,18178,9153,27850,21288,27274


In [47]:
# region that sells the most laptops 
df.pivot_table(values='Sales',columns='Product',index='Region',aggfunc='sum')['Laptop'].idxmax()

'East'

Question 8: Profitability Analysis

          Calculate the average profit for each product category.
          Which category is most profitable?

In [48]:
# Average profit per product category 
df.groupby('Category')['Profit'].mean()

Category
Accessories    410.603086
Electronics    396.462114
Office         471.938462
Name: Profit, dtype: float64

In [49]:
# Most profitable category 
df.groupby('Category')['Profit'].mean().idxmax()

'Office'

Question 9 :Discount Impact Study

          Analyze the relationship between discount and profit.
      
          Tasks:
            Calculate the average profit for each discount level
            Identify whether higher discounts lead to lower profit

In [50]:
# average profit for each discount level 
df.groupby('Discount')['Profit'].mean()

Discount
0.00    751.210588
0.05    531.368039
0.10    346.460513
0.15    237.191400
0.20     90.026923
Name: Profit, dtype: float64

In [52]:
# finding out whether higher discounts lead to lower profit
pd.crosstab(df['Discount'],df['Profit'])

Profit,-104.30,-77.55,-71.66,-67.01,-58.60,-49.47,-39.13,-36.79,-26.75,-25.47,...,1171.08,1245.37,1265.49,1266.10,1358.49,1422.83,1482.52,1495.12,1498.02,1578.32
Discount,,,,,,,,,,,,,,,,,,,,,
0.00,0,0,0,0,0,0,0,0,0,0,...,1,1,1,0,1,1,1,1,1,1
0.05,0,0,0,0,0,0,0,0,0,0,...,0,0,0,1,0,0,0,0,0,0
0.10,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
0.15,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
0.20,1,1,1,1,1,1,1,1,1,1,...,0,0,0,0,0,0,0,0,0,0


Profits are higher when discounts are lower.
Therefore, the business has to minimize on discounts to elevate profits the more.

Question 10: End-to-End Data Wrangling Task

           Perform the following workflow:
            Filter the dataset where Sales > 3000
            Group data by Region and Product
            Calculate total sales
            Reshape the result into a pivot table
           Explain the business insights obtained from this analysis.

In [54]:
# dataset where Sales > 3000 
df[df['Sales'] > 3000]

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30
5,1006,2024-07-04,Pooja,Bangalore,West,Printer,Office,3073,8,0.20,191.58
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
...,...,...,...,...,...,...,...,...,...,...,...
219,1220,2024-03-22,Meera,Bangalore,West,Phone,Electronics,5986,7,0.05,986.70
222,1223,2024-03-04,Meera,Ahmedabad,South,Phone,Electronics,4356,6,0.10,416.87
223,1224,2024-01-07,Dev,Bangalore,East,Tablet,Electronics,4115,4,0.05,870.03
225,1226,2024-06-14,Anita,Ahmedabad,East,Laptop,Electronics,5647,2,0.15,662.00


In [58]:
# Grouping data by Region and Product 
df.groupby(['Region','Product']).head()

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
0,1001,2024-10-01,Priya,Hyderabad,East,Mouse,Accessories,1735,7,0.00,307.52
1,1002,2024-05-03,Sneha,Kolkata,East,Monitor,Electronics,807,2,0.15,49.89
2,1003,2024-01-26,Anita,Mumbai,West,Phone,Electronics,2328,1,0.20,66.47
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30
...,...,...,...,...,...,...,...,...,...,...,...
210,1211,2024-07-21,Dev,Bangalore,North,Monitor,Electronics,2417,5,0.20,-58.60
214,1215,2024-08-25,Sneha,Pune,North,Laptop,Electronics,4540,6,0.00,940.54
218,1219,2024-02-16,Vikas,Hyderabad,West,Headphones,Accessories,1025,7,0.00,246.65
222,1223,2024-03-04,Meera,Ahmedabad,South,Phone,Electronics,4356,6,0.10,416.87


In [59]:
# total sales 
df['Sales'].sum()

np.int64(744220)

In [71]:
 #Reshaping the result into a pivot table 

# filtering Sales > 3000 
df_filtered = df[df['Sales']> 3000]

df_filtered.pivot_table(values='Sales',columns='Product',index='Region',aggfunc='sum')

Product,Headphones,Keyboard,Laptop,Monitor,Mouse,Phone,Printer,Tablet
Region,,,,,,,,
East,14469,9642,44096,16420,19000,29500,3696,29229
North,5326,15653,4540,5087,22209,24670,17185,3392
South,10026,19249,17950,14921,26928,18218,14143,19613
West,13898,33380,5922,14354,3511,15653,12648,17185


Business insight derived from this analysis: 

The sale of phone generates the most sales 
The sale of tablet generates the least sales 
The East Region has the best sales 